fs/background-fetcher/background_fetcher.go

```diff
@@ -0,0 +1,162 @@
+
+	// The minimum size of data to be fetched in a single fetch.
+	defaultMinFetchSize = 1 << 22 // 4 MiB
+
+	// Specifies how often the fetch will occur.
+	// The background fetcher will fetch a single span every `defaultFetchPeriod`.
+	defaultFetchPeriod = 500 * time.Millisecond
+)
+
+func WithMinFetchSize(size int64) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.minFetchSize = size
+		return nil
+	}
+}
+
+func WithSilencePeriod(period time.Duration) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.silencePeriod = period
+		return nil
+	}
+}
+
+func WithFetchPeriod(period time.Duration) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.fetchPeriod = period
+		return nil
+	}
+}
+
+// A backgroundFetcher is responsible for fetching spans from layers
+// in the background.
+type BackgroundFetcher struct {
+	minFetchSize  int64
+	silencePeriod time.Duration
+	fetchPeriod   time.Duration
+
+	rateLimiter *rate.Limiter
+
+	// All span managers are added to the channel and picked up in Run().
+	// If a span manager is still able to fetch, it is reinserted into the chanel.
+	waitQueue chan *layerResolver
+
+	closeChan chan struct{}
+
+	// The timer that the bg-fetcher will wait on until it is ready to (re)start
+	// background fetches.
+	timer *time.Timer
+}
+
+func NewBackgroundFetcher(opts ...Option) (*BackgroundFetcher, error) {
+	bf := newWithDefaults()
+	for _, o := range opts {
+		if err := o(bf); err != nil {
+			return nil, err
+		}
+	}
+	// Create a rate-limited that will fetch every bf.fetchPeriod
+	// with a burst capacity of 1 (i.e., it will never invoke more than 1 bg-fetch
+	// within bf.fetchPeriod)
+	bf.rateLimiter = rate.NewLimiter(rate.Every(bf.fetchPeriod), 1)
+	return bf, nil
+}
+
+func newWithDefaults() *BackgroundFetcher {
+	return &BackgroundFetcher{
+		minFetchSize:  defaultMinFetchSize,
+		silencePeriod: defaultSilencePeriod,
+		fetchPeriod:   defaultFetchPeriod,
+		waitQueue:     make(chan *layerResolver),
+		closeChan:     make(chan struct{}),
+	}
+}
+
+// Adds a new span manager to be background fetched from.
+// Sends the manager through the channel, which will be received in the Run() method.
+// Also resets the `timeTicker` so that background fetches are suspended until `silencePeriod`.
+func (bf *BackgroundFetcher) Add(dgst digest.Digest, mgr *sm.SpanManager) {
+	bf.waitQueue <- &layerResolver{SpanManager: mgr, layerDigest: dgst}
+	if bf.timer == nil {
+		bf.timer = time.NewTimer(bf.silencePeriod)
+	} else {
+		bf.timer.Reset(bf.silencePeriod)
+	}
+}
+
+func (bf *BackgroundFetcher) Close() error {
+	bf.closeChan <- struct{}{}
+	return nil
+}
+
+func (bf *BackgroundFetcher) Run(ctx context.Context) error {
+	for {
+
+		// If any new span managers have come in, wait for silencePeriod.
+		if bf.timer != nil {
+			<-bf.timer.C
+		}
+
+		if err := bf.rateLimiter.Wait(ctx); err != nil {
+			return fmt.Errorf("background fetch: error while waiting for rate limiter: %w", err)
+		}
+
+		select {
```

Comment:
You should check ctx.Done() as well to cancel this when the context is cancelled.

Reply:
Ack.

Reply:
Done.